File: backend/controllers/announcementController.js
const Announcement = require("../models/Announcement");
const User = require("../models/User");

// CREATE (caller must be Admin or Faculty — enforced in route)
exports.createAnnouncement = async (req, res) => {
  try {
    const dbUser = await User.findById(req.user.sub).select("role name email");
    if (!dbUser || (dbUser.role !== "Admin" && dbUser.role !== "Faculty")) {
      return res
        .status(403)
        .json({ message: "Only Admin or Faculty can post announcements" });
    }

    const { title, description, author, role } = req.body;

    const newAnnouncement = await Announcement.create({
      title,
      description,
      author: author || dbUser.name || dbUser.email,
      role: role || dbUser.role,
    });

    res.status(201).json({
      message: "Announcement created ✅",
      data: newAnnouncement
    });

  } catch (err) {
    res.status(500).json({ message: "Server error" });
  }
};

// GET ALL
exports.getAnnouncements = async (req, res) => {
  try {
    const data = await Announcement.find().sort({ createdAt: -1 });

    res.json(data);
  } catch (err) {
    res.status(500).json({ message: "Error fetching" });
  }
};

// DELETE (Admin only)
exports.deleteAnnouncement = async (req, res) => {
  try {
    const dbUser = await User.findById(req.user.sub).select("role");
    if (!dbUser || dbUser.role !== "Admin") {
      return res
        .status(403)
        .json({ message: "Only Admin can delete announcements" });
    }

    const { id } = req.params;
    const deleted = await Announcement.findByIdAndDelete(id);
    if (!deleted) {
      return res.status(404).json({ message: "Announcement not found" });
    }

    res.json({ message: "Announcement deleted" });
  } catch (err) {
    res.status(500).json({ message: "Server error" });
  }
};

File: DotNetBackend/Controllers/AnnouncementController.cs
using DotNetBackend.Models;
using DotNetBackend.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using System.Security.Claims;

namespace DotNetBackend.Controllers
{
    [ApiController]
    [Route("api/announcements")]
    [Authorize]
    public class AnnouncementController : ControllerBase
    {
        private readonly MongoDbService _dbService;

        public AnnouncementController(MongoDbService dbService)
        {
            _dbService = dbService;
        }

        [HttpPost]
        public async Task<IActionResult> CreateAnnouncement([FromBody] Announcement req)
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(userId)) return Unauthorized();

            var user = await _dbService.Users.Find(u => u.Id == userId).FirstOrDefaultAsync();
            if (user == null || (user.Role != "Admin" && user.Role != "Faculty"))
            {
                return StatusCode(403, new { message = "Only Admin or Faculty can post announcements" });
            }

            var announcement = new Announcement
            {
                Title = req.Title,
                Description = req.Description,
                Author = req.Author ?? (string.IsNullOrEmpty(user.Name) ? user.Email : user.Name),
                Role = req.Role ?? user.Role,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            await _dbService.Announcements.InsertOneAsync(announcement);

            return StatusCode(201, new { message = "Announcement created ✅", data = announcement });
        }

        [HttpGet]
        [AllowAnonymous]
        public async Task<IActionResult> GetAnnouncements()
        {
            var announcements = await _dbService.Announcements.Find(_ => true)
                .SortByDescending(a => a.CreatedAt)
                .ToListAsync();

            return Ok(announcements);
        }

        [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteAnnouncement(string id)
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            var user = await _dbService.Users.Find(u => u.Id == userId).FirstOrDefaultAsync();
            if (user == null || user.Role != "Admin")
            {
                return StatusCode(403, new { message = "Only Admin can delete announcements" });
            }

            var result = await _dbService.Announcements.DeleteOneAsync(a => a.Id == id);
            if (result.DeletedCount == 0)
            {
                return NotFound(new { message = "Announcement not found" });
            }

            return Ok(new { message = "Announcement deleted" });
        }
    }
}
